common/src/main/java/com/wynntils/screens/chattabs/widgets/ChatTabsWidget.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright © Wynntils 2023.
+ * This file is released under LGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.screens.chattabs.widgets;
+
+import com.mojang.blaze3d.vertex.PoseStack;
+import com.wynntils.core.components.Managers;
+import com.wynntils.core.components.Services;
+import com.wynntils.core.text.StyledText;
+import com.wynntils.screens.chattabs.ChatTabEditingScreen;
+import com.wynntils.services.chat.ChatTab;
+import com.wynntils.utils.colors.CommonColors;
+import com.wynntils.utils.mc.McUtils;
+import com.wynntils.utils.render.FontRenderer;
+import com.wynntils.utils.render.RenderUtils;
+import com.wynntils.utils.render.type.HorizontalAlignment;
+import com.wynntils.utils.render.type.TextShadow;
+import com.wynntils.utils.render.type.VerticalAlignment;
+import net.minecraft.ChatFormatting;
+import net.minecraft.client.gui.GuiGraphics;
+import net.minecraft.client.gui.components.AbstractWidget;
+import net.minecraft.client.gui.components.Button;
+import net.minecraft.client.gui.narration.NarrationElementOutput;
+import net.minecraft.network.chat.Component;
+
+public class ChatTabsWidget extends AbstractWidget {
+    private final float gridDivisions;
+    private final ChatTab chatTab;
+    private final Button deleteButton;
+    private final Button moveUpButton;
+    private final Button moveDownButton;
+    private final ChatTabEditingScreen parent;
+
+    public ChatTabsWidget(
+            float x,
+            float y,
+            int width,
+            int height,
+            ChatTab chatTab,
+            float gridDivisions,
+            ChatTabEditingScreen parent) {
+        super((int) x, (int) y, width, height, Component.literal(chatTab.getName()));
+
+        this.chatTab = chatTab;
+        this.gridDivisions = gridDivisions;
+        this.parent = parent;
+
+        this.deleteButton = new Button.Builder(
```

Comment:
instead of rendering and listening for them individually, you can add them as renderable widgets. Then all the events get called without you having to worry about that and they also get rendered (maybe you have to call super in `renderWidget` for that).

Reply:
You are in the widget itself..?

Reply:
Theoretically it could be passed up but we have never done it that way

Reply:
oh that's only available in screens, isn't it? ConfigTile does it like this too, but Minecrafts screen system isn't designed for that. Maybe we should start doing something about it like only using such classes only as builders for the actual widgets cuz in implementations like this stuff gets overlooked like forwarding the other events.

Reply:
it doesn't get overlooked if it is tested/coded properly? not every event needs to be passed down if the widget doesnt need it